ATS.html.markdown

```diff
@@ -0,0 +1,607 @@
+// val x: int = get_first (list_nil()) // doesn't compile
+
+// in the stdlib:
+// sortdef nat = {n:int | n >= 0}
+// sortdef pos = {n:int | n >= 1}
+
+fn {a:t@ype} also_get_first {n:pos} (xs: list(a, n)): a =
+  let
+    val+ list_cons(x, _) = xs // val+ also works
+  in x end
+
+// tail-recursive reverse
+fun {a:t@ype} reverse {n:int} (xs: list(a, n)): list(a, n) =
+  let
+    // local functions can use type variables from their enclosing scope
+    // this one uses both 'a' and 'n'
+    fun rev_helper {i:nat} (xs: list(a, n-i), acc: list(a, i)): list(a, n) =
+      case xs of
+      | list_nil() => acc
+      | list_cons(x, rest) => rev_helper(rest, list_cons(x, acc))
+  in rev_helper(xs, list_nil) end
+
+// ATS has three context-dependent namespaces
+// the two 'int's mean different things in this example, as do the two 'n's
+fn namespace_example {n:int} (n: int n): int n = n
+//                      ^^^                         sort namespace
+//                    ^          ^^^ ^   ^^^ ^      statics namespace
+// ^^^^^^^^^^^^^^^^^          ^                  ^  value namespace
+
+// a termination metric can go in .< >.
+// it must decrease on each recursive call
+// then ATS will prove it doesn't infinitely recurse
+fun terminating_factorial {n:nat} .<n>. (n: int n): int =
+  if n = 0 then 1 else n * terminating_factorial (n-1)
+
+
+/*************** Part 3: the LF fragment ****************/
+
+// ATS supports proving theorems in LF (http://twelf.org/wiki/LF)
+
+// Relations are represented by inductive types
+
+// Proofs that the nth fibonacci number is f
+dataprop Fib(n:int, m:int) =
+  | FibZero(0, 0)
+  | FibOne(1, 1)
+  | {n, f1, f2: int} FibInd(n, f1 + f2) of (Fib(n-1,f1), Fib(n-2,f2))
+
+// Proved-correct fibonacci implementation
+// [A] B is an existential type: "there exists A such that B"
+// (proof | value)
+fun fib {n:nat} .<n>. (n: int n): [f:int] (Fib(n,f) | int f) =
+  if n = 0 then (FibZero | 0) else
+  if n = 1 then (FibOne | 1) else
+  let
+    val (proof1 | val1) = fib (n-1)
+    val (proof2 | val2) = fib (n-2)
+  // the existential type is inferred
+  in (FibInd(proof1, proof2) | val1 + val2) end
+
+// Faster proved-correct fibonacci implementation
+fn fib_tail {n:nat} (n: int n): [f:int] (Fib(n,f) | int f) =
+  let
+    fun loop {i:int | i < n} {f1, f2: int} .<n - i>.
+          (p1: Fib(i,f1), p2: Fib(i+1,f2)
+          | i: int i, f1: int f1, f2: int f2, n: int n
+          ): [f:int] (Fib(n,f) | int f) =
+      if i = n - 1
+      then (p2 | f2)
+      else loop (p2, FibInd(p2,p1) | i+1, f2, f1+f2, n)
+  in if n = 0 then (FibZero | 0) else loop (FibZero, FibOne | 0, 0, 1, n) end
+
+// Proof-level lists of ints, of type 'sort'
+datasort IntList = ILNil of ()
+                 | ILCons of (int, IntList)
+
+// ILAppend(x,y,z) iff x ++ y = z
+dataprop ILAppend(IntList, IntList, IntList) =
+  | {x:IntList} AppendNil(x, ILNil, x)
+  | {a:int} {x,y,z: IntList}
+    AppendCons(x, ILCons(a,y), ILCons(a,z)) of ILAppend(x,y,z)
```

Comment:
Shouldn't this be `AppendCons(ILCons(a, x), y, ILCons(a,z)) of ILAppend(x,y,z)` for `x ++ y = z` to hold?
If so, the base clause seems to be flipped as well.

Reply:
You're absolutely right, I switched the arguments -- thanks for catching it.  I will fix it and rebase the PR.